zegami_sdk/source.py

```diff
@@ -51,3 +68,95 @@ def _imageset_dataset_join_id(self):
+        urls, id_set = client._obtain_signed_blob_storage_urls(self._collection.workspace_id, id_count=len(paths))
+
+        for path in paths:
+            file_name = os.path.basename(path)
+            file_ext = os.path.splitext(path)[-1]
+
+            # Check if the file has a correct extension
+            if mime_type:
+                mime_type = mime_type
+            elif file_ext in self.IMAGE_MIMES.keys():
+                mime_type = self.IMAGE_MIMES[file_ext]
+            else:
+                raise ValueError(
+                    f"File extension must be one of these: {' '.join([key for key in self.IMAGE_MIMES.keys()])}")
+
+            with open(path, 'rb') as f:
+
+                # find the blob storage url and upload to it
+                blob_id = id_set["ids"][index]
+                url = urls[blob_id]
+
+                # Make sure the upload isn't interputed for all images
+                try:
+                    client._upload_to_signed_blob_storage_url(f, url, mime_type)
+                except Exception as ex:
+                    print('\nAn exception occured: ', ex)
+
+                # update the new image details to the relevant endpoint
+                info = {
+                    "image": {
+                        "blob_id": blob_id,
+                        "name": file_name,
+                        "size": os.path.getsize(path),
+                        "mimetype": mime_type
+                    }
+                }
+
+                # append to the list in order to update later
+                images_info.append(info['image'])
+                index = index + 1
+
+        add_images_bulk_url = f'{client.HOME}/{client.API_0}/project/{self._collection.workspace_id}' \
+            f'/imagesets/{self._imageset_id}/images_bulk'
+
+        client._auth_post(add_images_bulk_url, body=None, return_response=True, json={'images': images_info})
+
+    def _assign_images_to_smaller_lists(self, image_dir):
+        """Create smaller lists based on the number of images in the directory."""
+        files = os.listdir(image_dir)
+        total_work = len(files)
+        final_paths = []
+        temp = []
+
+        if total_work > 2500:
+            size = 100
+        elif total_work < 100:
+            size = 1
+        else:
+            size = 10
+
+        i = 0
+        while i < total_work:
+            path = os.path.join(image_dir, files[i])
+            temp.append(path)
+            i += 1
+            if len(temp) == size or i == total_work:
+                final_paths.append(temp)
+                temp = []
+
+        return final_paths, total_work, size
+
+    def _upload_all_images(self, image_dir, mime_type=None, max_workers=50, show_time_taken=True):
+        """Upload multiple images to a collection."""
+        t = time()
+        all_paths, total_work, size = self._assign_images_to_smaller_lists(image_dir)
+        with tqdm(total=total_work) as pbar:
```

Comment:
I am not sure we want to use a progress bar within the SDK as it might not be run on the command line (not sure if this actually matters..?)

Reply:
I imagine it might matter if the SDK is a successor to the CLI and people use it interactively

Reply:
I think the thing using the SDK would normally be responsible for showing the progress bar. Typically in an SDK you would be able to pass in a callback to receive status updates.

That's probably where we want to go eventually, but I'm not too worried if it stays like this for now.